### PR TITLE
Use async/await instead of then

### PR DESCRIPTION
**Pull request by darkwing**
Much easier to read than the mess of `then`'s.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -344,34 +329,28 @@ class KeyringController extends EventEmitter {
...
+    } else {
+      throw new Error(
+        `Keyring ${keyring.type} doesn't support account removal operations`,
+      );
+    }
+
+    const accounts = await keyring.getAccounts();
+    // Check if this was the last/only account
+    if (accounts.length === 0) {
+      await this.removeEmptyKeyrings();
```
Comment (mcmire) [inline index.js:348]: Looks like we might want to say:

``` suggestion
      return await this.removeEmptyKeyrings();
```

File: index.js [hunk 2]
```diff
@@ -324,14 +313,10 @@ class KeyringController extends EventEmitter {
...
-      return this.getKeyringForAccount(address).then((keyring) => {
-        return keyring.exportAccount(normalizeAddress(address));
-      });
-    } catch (e) {
-      return Promise.reject(e);
-    }
+  async exportAccount(address) {
+    const keyring = await this.getKeyringForAccount(address);
+    const exported = keyring.exportAccount(normalizeAddress(address));
+    return exported;
```
Comment (mcmire) [inline index.js:319]: We should be able to shorten this to:

``` suggestion
    return await keyring.exportAccount(normalizeAddress(address));
```

File: index.js [hunk 3]
```diff
@@ -170,12 +167,10 @@ class KeyringController extends EventEmitter {
...
-  submitPassword(password) {
-    return this.unlockKeyrings(password).then((keyrings) => {
-      this.keyrings = keyrings;
-      this.setUnlocked();
-      return this.fullUpdate();
-    });
+  async submitPassword(password) {
+    this.keyrings = await this.unlockKeyrings(password);
+    this.setUnlocked();
+    this.fullUpdate(); // needs await?
```
Comment (mcmire) [inline index.js:173]: `fullUpdate` doesn't seem like an `async` method, so I don't think we need to await it.

File: index.js [hunk 4]
```diff
@@ -388,11 +367,10 @@ class KeyringController extends EventEmitter {
...
    * @returns {Promise<Object>} The signed transaction object.
    */
-  signTransaction(ethTx, _fromAddress, opts = {}) {
+  async signTransaction(ethTx, _fromAddress, opts = {}) {
     const fromAddress = normalizeAddress(_fromAddress);
-    return this.getKeyringForAccount(fromAddress).then((keyring) => {
-      return keyring.signTransaction(fromAddress, ethTx, opts);
-    });
+    const keyring = await this.getKeyringForAccount(fromAddress);
+    return keyring.signTransaction(fromAddress, ethTx, opts);
```
Comment (mcmire) [inline index.js:373]: Nit: Do we want to `await` this?

``` suggestion
    return await keyring.signTransaction(fromAddress, ethTx, opts);
```

File: index.js [hunk 5]
```diff
@@ -403,11 +381,10 @@ class KeyringController extends EventEmitter {
...
    * @returns {Promise<Buffer>} The raw signature.
    */
-  signMessage(msgParams, opts = {}) {
+  async signMessage(msgParams, opts = {}) {
     const address = normalizeAddress(msgParams.from);
-    return this.getKeyringForAccount(address).then((keyring) => {
-      return keyring.signMessage(address, msgParams.data, opts);
-    });
+    const keyring = await this.getKeyringForAccount(address);
+    return keyring.signMessage(address, msgParams.data, opts);
```
Comment (mcmire) [inline index.js:387]: Nit: Do we want to `await` this?

``` suggestion
    return await keyring.signMessage(address, msgParams.data, opts);
```

File: index.js [hunk 6]
```diff
@@ -419,11 +396,10 @@ class KeyringController extends EventEmitter {
...
    * @returns {Promise<Buffer>} The raw signature.
    */
-  signPersonalMessage(msgParams, opts = {}) {
+  async signPersonalMessage(msgParams, opts = {}) {
     const address = normalizeAddress(msgParams.from);
-    return this.getKeyringForAccount(address).then((keyring) => {
-      return keyring.signPersonalMessage(address, msgParams.data, opts);
-    });
+    const keyring = await this.getKeyringForAccount(address);
+    return keyring.signPersonalMessage(address, msgParams.data, opts);
```
Comment (mcmire) [inline index.js:402]: Nit: Do we want to `await` this? (and other examples)

File: index.js [hunk 7]
```diff
@@ -434,11 +410,10 @@ class KeyringController extends EventEmitter {
...
    * @returns {Promise<Buffer>} The public key.
    */
-  getEncryptionPublicKey(_address, opts = {}) {
+  async getEncryptionPublicKey(_address, opts = {}) {
     const address = normalizeAddress(_address);
-    return this.getKeyringForAccount(address).then((keyring) => {
-      return keyring.getEncryptionPublicKey(address, opts);
-    });
+    const keyring = await this.getKeyringForAccount(address);
+    return keyring.getEncryptionPublicKey(address, opts);
```
Comment (mcmire) [inline index.js:416]: Nit: Do we want to `await` this?

**Review by mcmire — COMMENTED** [review]
Good idea! Had some suggestions.

File: index.js [hunk 8]
```diff
@@ -97,7 +97,7 @@ class KeyringController extends EventEmitter {
    * either as a string or an array of UTF-8 bytes that represent the string.
    * @returns {Promise<Object>} A Promise that resolves to the state.
    */
-  createNewVaultAndRestore(password, seedPhrase) {
+  async createNewVaultAndRestore(password, seedPhrase) {
```
Comment (Gudahtt) [inline index.js:100]: Nit: In an `async` function it's more idiomatic to throw errors rather than return rejections, so we should probably update the two rejections below here as well

File: index.js [hunk 9]
```diff
@@ -692,35 +660,35 @@ class KeyringController extends EventEmitter {
    * @param {string} address - An account address.
```
Comment (Gudahtt) [inline index.js:660]: Nit: There is a `.then` a few lines up from here in `getAccounts` that we can remove as well

File: index.js [hunk 10]
```diff
@@ -546,15 +517,13 @@ class KeyringController extends EventEmitter {
...
-        new Error('KeyringController - password is not a string'),
-      );
+      throw new Error('KeyringController - password is not a string');
     }
 
     this.password = password;
-    return Promise.all(
+    const serializedKeyrings = await Promise.all(
       this.keyrings.map((keyring) => {
         return Promise.all([keyring.type, keyring.serialize()]).then(
```
Comment (Gudahtt) [inline index.js:528]: Nit: We can replace the `.then` here with an `await` as well

**Review by Gudahtt — COMMENTED** [review]
Great work @darkwing ! This has been on my wishlist for a long time. A few issues and suggestions but everything not commented on looks good

File: index.js (hunk 3 above)
Comment (darkwing) [reply index.js:173]: Good call -- sorry for leaving this in, was a TODO for me and I just forgot to remove it.

File: index.js (hunk 8 above)
Comment (darkwing) [reply index.js:100]: Good shout -- I did this elsewhere but not here.  Fixed.

File: index.js (hunk 10 above)
Comment (darkwing) [reply index.js]: Done!

**Review by ryanml — APPROVED** [review]
awesome!

**Review by Gudahtt — APPROVED** [review]
LGTM!
